[lldb] Use LLVM's helper for Unicode conversion (NFC)

JDevlieghere opened the pull request:
The codecvt header has been deprecated in C++17. Use LLVM's unicode
helpers to convert between UTF-8 and UTF-16.

JDevlieghere commented:
Looks like this is failing in CI. I'll take a look at this tomorrow. In the meantime I'd still like to merge #112446 and #112276.

labath commented on lldb/source/Host/common/Editline.cpp:
```diff
@@ -76,6 +76,46 @@ using namespace lldb_private::line_editor;
+  const wchar_t *in_next;
+  char *out_next;
+
+  if (cvt.out(mbs, in.data(), in.data() + length + 1, in_next, output.data(),
```
in.begin/end() ?

labath commented on lldb/source/Host/common/Editline.cpp:
```diff
@@ -76,6 +76,46 @@ using namespace lldb_private::line_editor;
+      std::use_facet<std::codecvt<wchar_t, char, std::mbstate_t>>(locale);
+
+  const size_t length = in.length();
+  std::string output(length + 1, 0x0);
```
This isn't right. A single wide character can take up to four bytes in the utf8 encoding. This works in `CompleteCharacter` as it's only interested in a single character (and it does the conversion in the other direction).  We might be able get away with overapproximating the size, given that the inputs here should be fairly small, but adding a loop with  partial conversions should not be that hard either.

labath commented on lldb/source/Host/common/Editline.cpp:
```diff
@@ -76,6 +76,46 @@ using namespace lldb_private::line_editor;
+      std::use_facet<std::codecvt<wchar_t, char, std::mbstate_t>>(locale);
+
+  const size_t length = in.length();
+  std::wstring output(length + 1, 0x0);
```
I believe this direction should be ok (even without the spurious +1).

labath reviewed (commented):
If you can't reproduce the buildbot errors, I can take a look, but I think the failures are likely due to the string length problems.

labath commented:
Actually (and, in retrospect, obviously) llvm already has [functions](https://github.com/llvm/llvm-project/blob/7dc2542ac24fcae89dfd179fa58c4ec4fb959e2b/llvm/include/llvm/Support/ConvertUTF.h#L227C6-L227C23) for wstring<->string conversions, so it would be best to use those.

JDevlieghere committed:
[lldb] Use LLVM's helper for Unicode conversion (NFC)

The codecvt header has been deprecated in C++17. Use LLVM's unicode
helpers to convert between UTF-8 and UTF-16.

JDevlieghere commented:
@labath FYI this is ready for review.

JDevlieghere commented:
Friendly ping

labath commented on lldb/source/Host/common/Editline.cpp:
```diff
@@ -1574,25 +1584,23 @@ bool Editline::CompleteCharacter(char ch, EditLineGetCharType &out) {
-    switch (cvt.in(state, input.begin(), input.end(), from_next, &out, &out + 1,
-                   to_next)) {
-    case std::codecvt_base::ok:
+    const char *cur_ptr = input.begin();
+    const char *end_ptr = input.end();
+    llvm::UTF32 code_point = 0;
+    llvm::ConversionResult cr = llvm::convertUTF8Sequence(
+        (const llvm::UTF8 **)&cur_ptr, (const llvm::UTF8 *)end_ptr, &code_point,
```
```suggestion
    auto *cur_ptr = reinterpret_cast<const llvm::UTF8 *>(input.begin());
    auto *end_ptr = reinterpret_cast<const llvm::UTF8 *>(input.end());
    llvm::UTF32 code_point = 0;
    llvm::ConversionResult cr = llvm::convertUTF8Sequence(
        &cur_ptr, end_ptr, &code_point,
```

I think this is "less undefined" :P

JDevlieghere committed:
Address Pavel's feedback
